rbs/RBSParser.cc

```diff
@@ -0,0 +1,73 @@
+#include "RBSParser.h"
+#include "core/errors/rewriter.h"
+
+namespace sorbet::rbs {
+
+std::unique_ptr<MethodType> RBSParser::parseSignature(core::Context ctx, Comment comment) {
+    rbs_string_t rbsString = {
+        .start = comment.string.data(),
+        .end = comment.string.data() + comment.string.size(),
+        .type = rbs_string_t::RBS_STRING_SHARED,
+    };
+
+    const rbs_encoding_t *encoding = &rbs_encodings[RBS_ENCODING_UTF_8];
+    lexstate *lexer = alloc_lexer(rbsString, encoding, 0, comment.string.size());
+    parserstate *parser = alloc_parser(lexer, 0, comment.string.size());
+
+    rbs_methodtype_t *rbsMethodType = nullptr;
+    parse_method_type(parser, &rbsMethodType);
+
+    if (parser->error) {
+        core::LocOffsets offset{
+            comment.loc.beginPos() + parser->error->token.range.start.char_pos + 2,
+            comment.loc.beginPos() + parser->error->token.range.end.char_pos + 2,
+        };
+
+        if (auto e = ctx.beginError(offset, core::errors::Rewriter::RBSError)) {
+            e.setHeader("Failed to parse RBS signature ({})", parser->error->message);
+        }
+
+        return nullptr;
+    }
+
+    free(parser);
+    free(lexer);
+
+    return std::make_unique<MethodType>(comment.loc, rbsMethodType);
+}
+
+std::unique_ptr<Type> RBSParser::parseType(core::Context ctx, Comment comment) {
+    rbs_string_t rbsString = {
+        .start = comment.string.data(),
+        .end = comment.string.data() + comment.string.size(),
+        .type = rbs_string_t::RBS_STRING_SHARED,
+    };
+
+    const rbs_encoding_t *encoding = &rbs_encodings[RBS_ENCODING_UTF_8];
+
+    lexstate *lexer = alloc_lexer(rbsString, encoding, 0, comment.string.size());
+    parserstate *parser = alloc_parser(lexer, 0, comment.string.size());
```

Comment:
Please use `unique_ptr<T, free>` for these sorts of things to avoid leaks along all paths out of the function (this function as currently written leaks in the case of an error from `parser`).

Also for members (e.g. in `Type`).

Reply:
I'm not sure what's the canonical way to handle free on C pointers, please let me know if this is not what you were expecting.

Reply:
If the value was allocated with `malloc`, and there's no other cleanup to do on the data held by the allocated value, then `free` would be the way to release that memory. Otherwise, if there's a more specific function that handles cleanup and calling `free`, you would use that one instead.

In this case, if there was a more specific `free_lexer` or `free_parser` function to use, those would be the right choice instead of `free`.